File: hdf-converters/converters/hdf-to-splunk/go/schema_validation_test.go
package hdftosplunk

import (
	"encoding/json"
	"fmt"
	"strings"
	"testing"

	corpus "github.com/mitre/hdf-libs/hdf-converters/v3/internal/corpus"
)

// NO PUBLISHED SCHEMA. Searched 2026-09-11: Splunk documents the HTTP Event
// Collector event envelope (event, time, host, source, sourcetype, index) in
// prose and publishes no JSON Schema for it; the payload under `event` is
// explicitly free-form, so there is nothing a document schema could constrain.
//
// What remains checkable is that every emitted line is a JSON object, which is
// what HEC requires and what a malformed serializer would break. A no-op
// validator would make every MustConvert contract pass vacuously.
type splunkNDJSONValidator struct{}

func (splunkNDJSONValidator) Validate(doc []byte) error {
	if len(doc) == 0 {
		return nil
	}
	for i, line := range strings.Split(strings.TrimSpace(string(doc)), "\n") {
		if line == "" {
			continue
		}
		var obj map[string]interface{}
		if err := json.Unmarshal([]byte(line), &obj); err != nil {
			return fmt.Errorf("line %d is not a JSON object: %w", i+1, err)
		}
	}
	return nil
}

// TestConvertHDFToSplunk_AdversarialCorpus holds this converter to the shared
// corpus contracts, with NDJSON well-formedness standing in for a schema.
func TestConvertHDFToSplunk_AdversarialCorpus(t *testing.T) {
	corpus.RunSchemaCorpus(t, splunkNDJSONValidator{}, corpus.ResultsCorpus(),
		func(in []byte) ([]byte, error) { return ConvertHDFToSplunk(in, "1.0.0") })
}

File: hdf-converters/converters/hdf-to-splunk/typescript/schema-validation.test.ts
import { describe, it, expect } from 'vitest';
import { resultsCorpus, runSchemaCorpus } from '../../../shared/typescript/schema-corpus.js';
import { convertHdfToSplunk } from './converter.js';

// NO PUBLISHED SCHEMA. Searched 2026-09-11: Splunk documents the HTTP Event Collector event envelope in prose and publishes no JSON Schema for it; the payload under `event` is explicitly free-form, so there is nothing a document schema could constrain.
//
// What remains checkable is that every emitted line is a JSON object, which the
// NDJSON contract requires. A no-op validator would make every MustConvert
// contract pass vacuously. Mirrors the Go peer.
const ndjsonObjects = (doc: unknown): string | null => {
  const text = typeof doc === 'string' ? doc : '';
  if (text === '') return null;
  const lines = text.trim().split('\n');
  for (const [i, line] of lines.entries()) {
    if (line === '') continue;
    let parsed: unknown;
    try {
      parsed = JSON.parse(line);
    } catch (e) {
      return `line ${i + 1} is not JSON: ${(e as Error).message}`;
    }
    if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
      return `line ${i + 1} is not a JSON object`;
    }
  }
  return null;
};

describe('Splunk adversarial corpus', () => {
  it('satisfies every contract for every case', async () => {
    const cases = resultsCorpus();
    expect(cases.length, 'an empty corpus would pass vacuously').toBeGreaterThan(0);
    await runSchemaCorpus(ndjsonObjects, cases, (input) => convertHdfToSplunk(input, '1.0.0'));
  });
});
